sql/log.cc

```diff
@@ -12934,3 +13110,227 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
+      cannot be renamed to binlog file.
+    - It is not supported to rename both statement cache and transaction cache
+      to binlog files at the same time.
+    - opt_optimize_thread_scheduling is just for testing purpose, it is usually
+      enabled, skip the disabled case to make the code simple.
+   */
+  if (encrypt_binlog || !opt_optimize_thread_scheduling ||
+      (entry->using_stmt_cache && entry->using_trx_cache &&
+       !stmt_cache->empty() && !trx_cache->empty()))
+    return false;
+
+  return true;
+}
+
+bool Binlog_commit_by_rotate::commit(MYSQL_BIN_LOG::group_commit_entry *entry)
+{
+  bool check_purge= false;
+  THD *thd= entry->thd;
+  binlog_cache_mngr *cache_mngr= entry->cache_mngr;
+  binlog_cache_data *cache_data= cache_mngr->get_binlog_cache_data(true);
+  if (unlikely(!entry->using_trx_cache || cache_data->empty()))
+    cache_data= cache_mngr->get_binlog_cache_data(false);
+
+  /* Call them before enter log_lock to avoid holding the lock long */
+  if (cache_data->sync_temp_file() || thd->wait_for_prior_commit())
+    return true;
+
+  // It will be released by trx_group_commit_leader
+  mysql_mutex_lock(&mysql_bin_log.LOCK_log);
+
+  enum enum_binlog_checksum_alg expected_alg=
+      mysql_bin_log.checksum_alg_reset != BINLOG_CHECKSUM_ALG_UNDEF
+          ? mysql_bin_log.checksum_alg_reset
+          : (enum_binlog_checksum_alg) binlog_checksum_options;
+
+  /*
+    In legacy mode, all events should has a valid position this done by
+    updating log_pos field when writing events from binlog cache to binlog
+    file. Thus rename binlog cache to binlog file is not supported in legacy
+    mode.
+
+    if the cache's checksum alg is not same to the binlog's checksum, it needs
+    to recalculate the checksum. Thus rename binlog cache to binlog file is
+    not supported.
+  */
+  if (!mysql_bin_log.is_open() || opt_binlog_legacy_event_pos ||
+      (expected_alg != cache_data->checksum_opt))
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    // It cannot do rename, so go to group commit
+    return mysql_bin_log.write_transaction_with_group_commit(entry);
+  }
+
+  m_entry= entry;
+  m_replaced= false;
+  m_cache_data= cache_data;
+  ulong prev_binlog_id= mysql_bin_log.current_binlog_id;
+
+  /*
+    Rotate will call replace_binlog_file() to rename the transaction's binlog
+    cache to the new binlog file.
+  */
+  if (mysql_bin_log.rotate(true, &check_purge, true /* commit_by_rotate */))
+  {
+    DBUG_ASSERT(!m_replaced);
+    DBUG_ASSERT(!mysql_bin_log.is_open());
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    return true;
+  }
+
+  if (!m_replaced)
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    if (check_purge)
+      mysql_bin_log.checkpoint_and_purge(prev_binlog_id);
+    // It cannot do rename, so go to group commit
+    return mysql_bin_log.write_transaction_with_group_commit(entry);
+  }
+
+  /* Seek binlog file to the end */
```

Comment:
I would place this 3 instruction block inside [Binlog_commit_by_rotate::replace_binlog_file](https://github.com/MariaDB/server/blob/39d134b19907af1368e037e97939aa2f3e6169d7/sql/log.cc#L13318) right after `m_replaced= true`.

Esp that it looks a bit worrying  that `detach_temp_file()` can change the size of the new active binlog file after `rotate()`  l.13204 is done. To me it'd be natural to invoke `my_chsize` closer to `m_replaced` sets to `true`. As to the worry, what if the server crashes right after `rotate()`? I imagine (from [here])(https://github.com/MariaDB/server/blob/39d134b19907af1368e037e97939aa2f3e6169d7/sql/log_cache.cc#L50) the renamed file may have some garbage at its tail part.

Reply:
Good point, thanks! Originally, detach_temp_file() happened before replace_bnlog_file(), so I put my_chsize() in it.
Now, I would put my_chsize() into Binlog_commit_by_rotate::commit() right before acquired LOCK_log. my_chsize() may take long time if it shrinks large space.

Reply:
The 3 line placed here, because the first line cannot be done in replace_binlog_file(),
```  
reinit_io_cache(&mysql_bin_log.log_file, WRITE_CACHE,
                  m_cache_data->temp_file_length(), false, true);" 
```
So I will put my_chsize just into Binlog_commit_by_rotate::commit() just before acquire LOCK_log.

Reply:
I also added a test case to verify the behavior.